File: front/src/domain/course.ts
import { Method } from "./method";
import { Room } from "./room";
import { Schedule } from "./schedule";

export type Course = {
  id: string;
  year: number;
  code: string;
  name: string;
  instructors: string[];
  credit: number;
  overview: string;
  remarks: string;
  recommendedGrades: number[];
  methods: Method[];
  schedules: Schedule[];
  rooms: Room[];
};

export type RegisteredCourse = {
  id: string;
  year: number;
  code?: string; // If the course is added by manual, code is null.
  name: string;
  instructors: string[];
  credit: number;
  methods: Method[];
  schedules: Schedule[];
  rooms: Room[];
  memo: string;
  attendance: number;
  absence: number;
  late: number;
  tagIds: string[];
};

File: back/handler/api/v4/rpc/timetable/v1/conv/course.go
package timetablev1conv

import (
	sharedconv "github.com/twin-te/twin-te/back/handler/api/v4/rpc/shared/conv"
	timetablev1 "github.com/twin-te/twin-te/back/handler/api/v4/rpcgen/timetable/v1"

	"github.com/twin-te/twin-te/back/base"
	timetabledomain "github.com/twin-te/twin-te/back/module/timetable/domain"
)

func ToPBCourse(course *timetabledomain.Course) (pbCourse *timetablev1.Course, err error) {
	pbCourse = &timetablev1.Course{
		Id:                sharedconv.ToPBUUID(course.ID),
		Year:              sharedconv.ToPBAcademicYear(course.Year),
		Code:              course.Code.String(),
		Name:              course.Name.String(),
		Instructors:       course.Instructors,
		Credit:            course.Credit.String(),
		Overview:          course.Overview,
		Remarks:           course.Remarks,
		LastUpdatedAt:     sharedconv.ToPBRFC3339DateTime(course.LastUpdatedAt),
		HasParseError:     course.HasParseError,
		IsAnnual:          course.IsAnnual,
		RecommendedGrades: base.Map(course.RecommendedGrades, ToPBRecommendedGrade),
	}

	pbCourse.Methods, err = base.MapWithErr(course.Methods, ToPBCourseMethod)
	if err != nil {
		return
	}

	pbCourse.Schedules, err = base.MapWithErr(course.Schedules, ToPBSchedule)
	if err != nil {
		return
	}

	return
}
